lib/util/state.js

```diff
@@ -25,9 +37,134 @@ export function getActiveItineraries (otpState) {
+      if (res && res.plan) itineraries.push(...res.plan.itineraries)
+    })
+  }
+  const {filter, sort} = otpState.filter
+  const {direction, type} = sort
+  return itineraries
+    .filter(itinerary => {
+      switch (filter) {
+        case 'ALL':
+          return true
+        case 'ACTIVE':
+          // ACTIVE filter checks that the only modes contained in the itinerary
+          // are 'active' (i.e., walk or bike). FIXME: add micromobility?
+          let allActiveModes = true
+          itinerary.legs.forEach(leg => {
+            if (!isWalk(leg.mode) && !hasBike(leg.mode)) allActiveModes = false
+          })
+          return allActiveModes
+        case 'TRANSIT':
+          return itineraryHasTransit(itinerary)
+        case 'CAR':
+          let hasCar = false
+          itinerary.legs.forEach(leg => {
+            if (isCar(leg.mode)) hasCar = true
+          })
+          return hasCar
+        default:
+          console.warn(`Filter (${filter}) not supported`)
+          return true
+      }
+    })
+    .sort((a, b) => {
+      switch (type) {
+        case 'WALKTIME':
+          if (direction === 'ASC') return a.walkTime - b.walkTime
+          else return b.walkTime - a.walkTime
+        case 'ARRIVALTIME':
+          if (direction === 'ASC') return a.endTime - b.endTime
+          else return b.endTime - a.endTime
+        case 'DEPARTURETIME':
+          if (direction === 'ASC') return a.startTime - b.startTime
+          else return b.startTime - a.startTime
+        case 'DURATION':
+          if (direction === 'ASC') return a.duration - b.duration
+          else return b.duration - a.duration
+        case 'COST':
+          const aTotal = getTotalFare(a)
+          const bTotal = getTotalFare(b)
+          if (direction === 'ASC') return aTotal - bTotal
+          else return bTotal - aTotal
+        default:
+          if (type !== 'BEST') console.warn(`Sort (${type}) not supported. Defaulting to BEST.`)
+          // FIXME: Fully implement default sort algorithm.
+          const aCost = calculateItineraryCost(a)
+          const bCost = calculateItineraryCost(b)
+          if (direction === 'ASC') return aCost - bCost
+          else return bCost - aCost
+      }
+    })
+}
+
+/**
+ * This calculates the "cost" (not the monetary cost, but the cost according to
+ * multiple factors like duration, total fare, and walking distance) for a
+ * particular itinerary, for use in sorting itineraries.
+ * FIXME: Do major testing to get this right.
+ */
+function calculateItineraryCost (itinerary) {
+  const FARE_COST_FACTOR = 0.5
+  const DURATION_COST_FACTOR = 0.25
+  const WALK_COST_FACTOR = 0.1
+  const WAIT_COST_FACTOR = 0.1
+  const TRANSFER_COST_FACTOR = 0.9
+  return getTotalFare(itinerary) * FARE_COST_FACTOR +
+    itinerary.duration * DURATION_COST_FACTOR +
+    itinerary.walkDistance * WALK_COST_FACTOR +
+    itinerary.waitingTime * WAIT_COST_FACTOR +
+    itinerary.transfers * TRANSFER_COST_FACTOR
```

Comment:
I think it'd be useful to include a drive reluctance factor in here. I think most of our customers will be transit agencies and will therefore not want itineraries with lots of driving to show up so high in the results.